dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -288,154 +278,408 @@ def read_metadata(
+        dtypes,
+        base_path,
+        paths,
+    ):
+        """Organize row-groups by file."""
+
+        # Get partitioning metadata
+        pqpartitions = pf.info.get("partitions", None)
+
+        # Get the number of row groups per file
+        single_rg_parts = int(split_row_groups) == 1
+        global_lookup = {}
+        file_row_groups = defaultdict(list)
+        file_row_group_stats = defaultdict(list)
+        file_row_group_column_stats = defaultdict(list)
+        cmax_last = {}
+        for rg, row_group in enumerate(pf.row_groups):
+
+            # We can filter partition columns here without dealing
+            # with statistics
+            if (
+                pqpartitions
+                and filters
+                and fastparquet.api.filter_out_cats(row_group, filters)
             ):
-                # Case (3B)
-                # We can pass a specific file/part path
-                path_curr = part.columns[0].file_path
-                if path_last and path_curr == path_last:
-                    i_path += 1
+                continue
+
+            fpath = row_group.columns[0].file_path
+            if fpath is None:
+                if paths and pf.fn in paths:
+                    # There doesn't need to be a file_path
+                    # in all cases for Fastparquet
+                    fpath = pf.fn
+                    base_path = base_path or ""
                 else:
-                    i_path = 0
-                    path_last = path_curr
-                file_path = base_path + path_curr
+                    raise ValueError(
+                        "Global metadata structure is missing a file_path string. "
+                        "If the dataset includes a _metadata file, that file may "
+                        "have one or more missing file_path fields."
+                    )
+
+            if file_row_groups[fpath]:
+                file_row_groups[fpath].append(file_row_groups[fpath][-1] + 1)
             else:
-                # Case (3C)
-                # We cannot pass a specific file/part path
-                file_path = paths
-                i_path = i
-
-            # Strip down pf object
-            if pf_deps and pf_deps != "tuple":
-                # Case (2)
-                for col in part.columns:
-                    col.meta_data.statistics = None
-                    col.meta_data.encoding_stats = None
-
-            # Final definition of "piece" and "pf" for this output partition
-            piece = i_path if pf_deps else part
-            pf_piece = (file_path, gather_statistics) if pf_deps == "tuple" else pf_deps
-            part_item = {
-                "piece": piece,
-                "kwargs": {"pf": pf_piece},
-            }
-            parts.append(part_item)
+                file_row_groups[fpath].append(0)
+
+            # Store mapping of local row-group to global row_group
+            global_lookup[(fpath, file_row_groups[fpath][-1])] = rg
```

Comment:
It's a bit messy...
There was a reason fastparquet didn't do this! :)

Reply:
Right - I'm not really sure of the best way to get what we want without having some sort of map between the local file row-groups and global row-group count.

Reply:
Is this concern a blocker for this PR? Either way, do you have thoughts/suggestions on a better way to do this?

Reply:
Note that the use of `global_lookup` is now removed - We are now storing the global row-group index within `file_row_groups` itself.